### PR TITLE
fix(jexl): separate extractor for mapby arguments

### PR DESCRIPTION
**Pull request by sbor23**
This fixes a regression, when table questions are referenced that haven't been answered, so no row document is available. In this case the jexl failed because no field exists for such a question.
Thus we separate the extractors and extract the mapby arguments only where need, that is to determine the calc dependents.

Theoretically it would still be correct to treat the arguments as fields too (if the subject is an answer transform), but then we would need a way to distinguish between actually missing questions (don't exist in the form but are referenced in the jexl) and missing table answers (that are "optional" because a table doesn't necessarily need to have rows).
With the separate extractor we could do that, but there is currently no need for it during regular validation steps.

### CODE REVIEW

**Review by anehx — DISMISSED** [review]

File: caluma/caluma_form/signals.py [hunk 1]
```diff
@@ -80,28 +86,24 @@ def _update_or_create_calc_answer(question, document):
...
+    # be invalid, in which case we return None
+    value = jexl.evaluate(field.question.calc_expression, raise_on_error=False)
+
+    try:
+        ans = models.Answer.objects.get(question=question, document=field.document)
+        update_model(ans, {"value": value})
+    except models.Answer.DoesNotExist:
+        models.Answer.objects.create(
+            question=question, document=field.document, value=value
+        )
```
[the remark this answers is not in the document]
Comment (sbor23) [reply caluma/caluma_form/signals.py:106]: The update wouldn't trigger historical entries this way...

[the remark this answers is not in the document]
Comment (anehx) [reply caluma/caluma_form/signals.py:106]: Not quite sure if that's true: https://docs.djangoproject.com/en/2.2/ref/models/querysets/#update-or-create it's a shortcut for

```python
defaults = {'first_name': 'Bob'}
try:
    obj = Person.objects.get(first_name='John', last_name='Lennon')
    for key, value in defaults.items():
        setattr(obj, key, value)
    obj.save()
except Person.DoesNotExist:
    new_values = {'first_name': 'John', 'last_name': 'Lennon'}
    new_values.update(defaults)
    obj = Person(**new_values)
    obj.save()
```

Which would trigger the historical entries.. You're confusing it with `.update()`

File: caluma/caluma_core/jexl.py [hunk 2]
```diff
@@ -99,7 +99,7 @@ def evaluate(self, expression, context=None):
             self._expr_stack.pop()
 
 
-class ExtractTransformReferenceAnalyzer(ValidatingAnalyzer):
+class ExtractTransformSubjectAnalyzer(ValidatingAnalyzer):
     """
     Extract all referenced values of given transforms, be it arguments or subjects.
```
[the remark this answers is not in the document]
Comment (sbor23) [reply caluma/caluma_core/jexl.py]: done

**Comment by anehx** [comment]
Looks good! How well was this tested?

**Comment by sbor23** [comment]
> Looks good! How well was this tested?

Pretty well from what I can tell, validations etc. worked as expected

**Comment by anehx** [comment]
> > Looks good! How well was this tested?
> 
> Pretty well from what I can tell, validations etc. worked as expected

In context of a project using caluma?

**Comment by sbor23** [comment]
> > > Looks good! How well was this tested?
> > 
> > 
> > Pretty well from what I can tell, validations etc. worked as expected
> 
> In context of a project using caluma?

Yes! Submit, pdf generation, manipulating the workflow etc.

**Pull request merged**